File: src/main/java/com/github/wakingrufus/mastodon/config/FileConfig.java
package com.github.wakingrufus.mastodon.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.github.wakingrufus.mastodon.account.AccountConfig;
import lombok.extern.slf4j.Slf4j;

import java.io.File;
import java.io.IOException;

@Slf4j
public class FileConfig implements Config {
    private final File file;
    private final ObjectMapper objectMapper;

    public FileConfig(File file, ObjectMapper objectMapper) {
        this.file = file;
        this.objectMapper = objectMapper;
    }


    @Override
    public ConfigData getConfig() {

        ConfigData configData = null;
        if (!file.exists()) {
            log.info("config not found, creating");
            try {
                boolean created = file.createNewFile();
                configData = new ConfigData();
            } catch (IOException e) {
                log.error("Error creating config file: " + e.getLocalizedMessage(), e);
            }
        } else {
            try {
                configData = objectMapper.readValue(file, ConfigData.class);
            } catch (IOException e) {
                log.error("Error reading config file: " + e.getLocalizedMessage(), e);
                configData = new ConfigData();
            }
        }
        return configData;
    }

    public void addAccount(AccountConfig identity) {
        ConfigData configData = getConfig();
        configData.getIdentities().add(identity);
        try {
            objectMapper.writeValue(file, configData);
        } catch (IOException e) {
            log.error("Error writing config file: " + e.getLocalizedMessage(), e);
        }
    }

}

File: src/main/kotlin/com/github/wakingrufus/mastodon/config/FileConfig.kt
package com.github.wakingrufus.mastodon.config

import com.fasterxml.jackson.databind.ObjectMapper
import com.github.wakingrufus.mastodon.account.AccountConfig
import mu.KLogging
import java.io.File
import java.io.IOException


class FileConfig(private val file: File, private val objectMapper: ObjectMapper) : Config {
    companion object : KLogging()


    override val config: ConfigData
        get() {

            var configData: ConfigData? = null
            if (!file.exists()) {
                logger.info("config not found, creating")
                try {
                    val created = file.createNewFile()
                    configData = ConfigData()
                } catch (e: IOException) {
                    logger.error("Error creating config file: " + e.localizedMessage, e)
                }

            } else {
                try {
                    configData = objectMapper.readValue(file, ConfigData::class.java)
                } catch (e: IOException) {
                    logger.error("Error reading config file: " + e.localizedMessage, e)
                    configData = ConfigData()
                }
            }
            return configData!!
        }

    override fun addAccount(identity: AccountConfig) {
        try {
            objectMapper.writeValue(file, config.copy(identities = config.identities.plus(identity)))
        } catch (e: IOException) {
            logger.error("Error writing config file: " + e.localizedMessage, e)
        }

    }

}
